parser/src/toMachineConfig.ts

```diff
@@ -0,0 +1,230 @@
+  }
+  if (astResult.onEntry) {
+    config.onEntry = getActionConfig(astResult.onEntry, opts);
+  }
+  if (astResult.exit) {
+    config.exit = getActionConfig(astResult.exit, opts);
+  }
+  if (astResult.onExit) {
+    config.onExit = getActionConfig(astResult.onExit, opts);
+  }
+
+  if (astResult.on) {
+    config.on = {};
+
+    astResult.on.properties.forEach((onProperty) => {
+      (config.on as any)[onProperty.key] = getTransitions(
+        onProperty.result,
+        opts,
+      );
+    });
+  }
+
+  if (astResult.after) {
+    config.after = {};
+
+    astResult.after.properties.forEach((afterProperty) => {
+      (config.after as any)[afterProperty.key] = getTransitions(
+        afterProperty.result,
+        opts,
+      );
+    });
+  }
+
+  if (astResult.history) {
+    config.history = astResult.history.value;
+  }
+
+  if (astResult.states) {
+    const states: typeof config.states = {};
+
+    astResult.states.properties.forEach((state) => {
+      states[state.key] = parseStateNode(state.result);
+    });
+
+    config.states = states;
+  }
+
+  if (astResult.always) {
+    config.always = getTransitions(astResult.always, opts);
+  }
+
+  if (astResult.meta?.description) {
+    config.meta = {
+      description: astResult.meta.description.value,
+    };
+  }
+
+  if (astResult.onDone) {
+    // @ts-ignore
+    config.onDone = getTransitions(astResult.onDone);
+  }
+
+  if (astResult.description) {
+    config.description = astResult.description.value;
+  }
+
+  if (astResult.invoke) {
+    const invokes: typeof config.invoke = [];
+
+    astResult.invoke.forEach((invoke) => {
+      if (!invoke.src) return;
+      let src: string;
+      if (opts?.hashInlineImplementations) {
+        src =
+          invoke.src.declarationType === "named"
+            ? invoke.src.value
+            : invoke.src.inlineDeclarationId;
+      } else {
+        src = invoke.src.value;
+      }
+      const toPush: typeof invokes[number] = {
+        src,
+      };
+
+      if (invoke.id) {
+        toPush.id = invoke.id.value;
+      }
+
+      if (invoke.autoForward) {
+        toPush.autoForward = invoke.autoForward.value;
+      }
+
+      if (invoke.forward) {
+        toPush.forward = invoke.forward.value;
+      }
+
+      if (invoke.onDone) {
+        // @ts-ignore
+        toPush.onDone = getTransitions(invoke.onDone);
+      }
+
+      if (invoke.onError) {
+        // @ts-ignore
+        toPush.onError = getTransitions(invoke.onError);
+      }
+
+      invokes.push(toPush);
+    });
+
+    if (invokes.length === 1) {
+      config.invoke = invokes[0];
+    } else {
+      config.invoke = invokes;
+    }
+  }
+
+  return config;
+};
+
+export const toMachineConfig = (
+  result: TMachineCallExpression,
+  opts?: ToMachineConfigParseOptions,
+): MachineConfig<any, any, any> | undefined => {
+  if (!result?.definition) return undefined;
+  return parseStateNode(result?.definition, opts);
```

Comment:
q: what the intermediate representation is good for? why don't we create the config while visiting the AST?

Reply:
For traversing the AST and holding meta information about how it's defined. It also exposes many helper methods for easier introspection/traversal. See `MachineParseResult.ts`